### PR TITLE
Adapt 3D-neuro-seg network for the deformations project

### PR DESCRIPTION
**Pull request by kdiers**
This PR adapts the original 3D-neuro-seg architecture for the predicition of deformations from T1 images. The goal is to use the RCVNet model and its VNet submodel to work with conformed T1 image inputs and corresponding deformation fields.

This PR pulls from the dev into the main branch of a forked version (3D-deformations) of the original 3D-neuro-seg repository. It is not intended to merge this back into the original repository, but rather keep the 3D-deformations repository as a separate project.

Specifically, we:

- added generate_hdf5.py
- updated utils/dataloader.py
    - removed weights
    - changed aseg to deformation field
- adapted models
    - removed unused code from models/RCVNet.py
    - removed QuadNet, CropNet3d, MultiResVNet models from model folder, but kept RCVNet submodels (even though currently only VNet is used)
- adapted utils/run_utils.py
    - removed unused code (quadnet, multiloss)
    - removed weights
    - changed aseg to deformation field (includes change of data type)
- adapted utils/losses.py
    - changed combined CE+Dice loss to MSE loss
- adapted train_model.py
    - changed parameters (num_classes: 3, model_name: RCVNet, sub_model_name: VNet)
    - removed unsused code since we only run RCVNet models
    - adjusted loss (no more class-wise losses, changed decision rules for 'trying again')
- adapted evaluation scripts
    - removed class-wise evaluation
    - changed Dice to MSE 
    - changed patch-based to full-volume evaluation
    - removed FastSurfer-based post-processing and evaluation
- updated requirements
- created dockerfiles
- introduced cmdline args

### CODE REVIEW

File: model/RCVNet.py [hunk 1]
```diff
@@ -74,16 +74,9 @@ def __init__(self, params):
...
-        # i.e. registered as a parameter, but never used
-        self.down_input_lower = nn.Sequential(
-                nn.Conv3d(in_channels=params['in_channels'], out_channels=4*params['out_channels'],
-                          kernel_size=(4,4,4), padding=0, stride=4),
-                nn.GroupNorm(num_groups=4, num_channels=4*params['out_channels']),
-                nn.PReLU()
-            )
-
         # Start model creation
+        self.interpolation_block_down = InterpolationBlock(params)
```
Comment (dkuegler) [inline model/RCVNet.py:78]: It looks a bit like you are directly interpolating the image, not the latent space.
Can you confirm you have a 3d Convolution Block before the interpolation block?

File: model/VNet.py [hunk 2]
```diff
@@ -72,18 +72,35 @@ def forward(self, x):
...
 
+class InterpolationBlock(nn.Module):
+    """
+    """
+
+    def __init__(self, params):
+
+        super(InterpolationBlock, self).__init__()
+
+        self.interpolation_block = nn.Upsample(scale_factor=params['sampling'], mode='trilinear')
```
Comment (dkuegler) [inline model/VNet.py:83]: I am not really sure why you are not using the FastSurferVINN interpolation block here.
There would be 2 reasons here:
1. you can specify different "samplings" for different patches (in the same batch)
2. you can feed our module different resolution images (say one 0.8 one 1.0 mm image).
The latter might not be as important as you are probably only doing 1.0mm images...
`torch.nn.Upsample(size=None, scale_factor=None, mode='nearest', align_corners=None, recompute_scale_factor=None)`

github.com/Deep-MI/FastSurfer/FastSurferCNN/models/interpolation_layer.py

File: model/RCVNet.py [hunk 3]
```diff
@@ -230,6 +235,9 @@ def eval_forward(self, x):
         x_last = multi_gpu_check(self.gpu_map, 'output_block', x_last)
         out = self.output_block(x_last)
 
+        # Interpolate
+        out = self.interpolation_block_up(out)
```
Comment (dkuegler) [inline model/RCVNet.py:239]: It also seems a bit like you have no skip connection on the native image level.

File: model/RCVNet.py [hunk 4]
```diff
@@ -230,6 +235,9 @@ def eval_forward(self, x):
         x_last = multi_gpu_check(self.gpu_map, 'output_block', x_last)
         out = self.output_block(x_last)
```
Comment (dkuegler) [inline model/RCVNet.py:236]: It seems you kept this the same, so you are also not resampling the latent space back, but the deformation field. While that might work, it is not really the idea for VINN, i.e. this is not really a VINN network.

File: model/RCVNet.py [hunk 5]
```diff
@@ -246,12 +254,6 @@ def __init__(self, params):
...
         self.gen_random = params['gen_random']
 
-        self.down_input_lower = nn.Sequential(
-            nn.Conv3d(in_channels=params['in_channels'], out_channels=4 * params['out_channels'],
-                      kernel_size=(4, 4, 4), padding=0, stride=4),
-            nn.GroupNorm(num_groups=4, num_channels=4 * params['out_channels']),
-            nn.PReLU()
-        )
         # in_channels: 16, out_channels: 16
         self.encoder_block_1 = AttEncoderBlock(params)
```
Comment (dkuegler) [inline model/RCVNet.py:258]: RCVNetAttention is not used i think, just the standard RCVNet.

**Review by dkuegler — COMMENTED** [review]
The VINN architecture is not really implemented as a VINN, but rather as a image-resample and predicted deformations field resample.

**Comment by dkuegler** [comment]
I dont particularly have any more comments.
Just that I would (also) do a script that not just predicts the Warp, but actually applies the warp so you directly get the deformed T1.
